test/src/plugin.test.js

```diff
@@ -0,0 +1,132 @@
+      expect(typeof plugins.length.afterMapping).toBe("function");
+      expect(plugins.length.beforeMapping).toBeUndefined();
+    });
+  });
+
+  describe("When register a valid plugin", () => {
+    const plugin = {
+      id: "test_plugin",
+      beforeMapping: (data) => data,
+      afterMapping: (data) => data,
+    };
+    const plugins = registerPlugins([plugin]);
+    it("Should return correct registered plugins", () => {
+      expect(Object.keys(plugins)).toContain("test_plugin");
+      expect(typeof plugins.test_plugin.beforeMapping).toBe("function");
+      expect(typeof plugins.test_plugin.afterMapping).toBe("function");
+    });
+  });
+
+  describe("When register an plugin without id", () => {
+    const plugin = {
+      beforeMapping: (data) => data,
+      afterMapping: (data) => data,
+    };
+
+    it("Should not register invalid correct and return correct registered plugins", () => {
+      const error = spyOn(global.console, "error");
+      registerPlugins([plugin]);
+
+      expect(error).toHaveBeenCalledWith("Register plugin error at index: 0. id is required.");
+    });
+  });
+
+  describe("When register an plugin without hook functions", () => {
+    const plugin = {
+      id: "invalid plugin",
+    };
+
+    it("Should not register invalid correct and return correct registered plugins", () => {
+      const error = spyOn(global.console, "error");
+      registerPlugins([plugin]);
+
+      expect(error).toHaveBeenCalledWith(
+        "Register plugin error at index: 0. id: \"invalid plugin\". Plugin's structure is not correct."
+      );
+    });
+  });
+});
+
+describe("Test for usePlugins", () => {
+  let plugins = null;
+  beforeAll(() => {
+    const cusPlugins = [
+      {
+        id: "plugin1",
+        beforeMapping: (data) => {
+          console.log("plugin 1 before");
+          return data;
+        },
+        afterMapping: (data) => {
+          console.log("plugin 1 after");
+          return data;
+        },
+      },
+      {
+        id: "plugin2",
+        beforeMapping: (data) => {
+          console.log("plugin 2 before");
+          return data;
+        },
+        afterMapping: (data) => {
+          console.log("plugin 2 after");
+          return data;
+        },
+      },
+    ];
+
+    plugins = registerPlugins(cusPlugins);
+  });
+
+  describe("When plugin id can be found", () => {
+    it("Should get correct callbacks", () => {
+      const itemPlugins = usePlugins(plugins, "plugin1");
+
+      const log = spyOn(global.console, "log");
+      itemPlugins.beforeMapping("test1");
+      expect(log).toHaveBeenCalledWith("plugin 1 before");
+
+      itemPlugins.afterMapping("test1");
+      expect(log).toHaveBeenCalledWith("plugin 1 after");
+    });
+
+    it("Should invoke callbacks in correct order", () => {
+      const itemPlugins = usePlugins(plugins, "plugin2|plugin1");
+
+      const log = spyOn(global.console, "log");
+      itemPlugins.beforeMapping("test1");
```

Comment:
❓ Is this supposed to call the plugin beforeMapping hook again? Should you test for `expect(log).toHaveBeenNthCalledWith(3, ...);`?

Reply:
The `spyOn` will reset the counter each time when it is called.